Init xk6 Docker plugin

lxkuz opened the pull request:
The update includes the next docker methods:

- docker ps
- docker start
- docker stop
- docker pause
- docker unpause
- docker inspect
- docker logs
- docker exec

lxkuz committed:
Add go project with k6 extention blank

lxkuz committed:
Fix extension name

lxkuz committed:
Add Readme

lxkuz committed:
Add Docker ps support

dgzlopes commented on Readme.md:
Nit: The script name it's not the same (test.js->script.js).

dgzlopes commented on docker.go:
```diff
@@ -0,0 +1,44 @@
+func (d *Docker) Ps() []string {
+	cli, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		panic(err)
```
I think it's better to return the error. But probably @simskij knows better :)

simonaronsson replied:
Yes, I agree that we should return the error here rather than panicking. The docker client erroring should probably not be considered fatal for the whole VU goroutine. 🤔

lxkuz replied:
✅

dgzlopes commented on docker.go:
```diff
@@ -0,0 +1,44 @@
+	result := make([]string, 0, len(containers))
+
+	for _, container := range containers {
+		result = append(result, fmt.Sprintf("%s %s %s", container.ID[:10], container.Image, container.Command))
```
Perhaps it's better to return a struct, instead of a string with the concatenated values.

From my POV it's easier to consume from JS.

simonaronsson replied:
I agree with @dgzlopes on this one. If the struct itself is too big/contains a lot of properties we don't care about, I'd suggest remapping it to a bespoke struct with reasonable properties and proper json tags.

Consider a scenario where I want to list my containers and get the command used to launch one named `whatever-nginx`. I'd then. have to regex parse the string from the JS runtime, which likely won't be that fast. If I got a json object on the other hand, it wouldn't be a big deal if I had to `console.log` the object myself.

lxkuz replied:
@simskij  please take a look at [rework commit](https://github.com/k6io/xk6-docker/pull/1/commits/1b5d376742d53ab533cc510c60dc957eaee9d56e). I tried to keep the initial interface that came from docker go lib. Example js file shows how to get normal `docker ps` output now. IMO the benefits of this approach are transparent support of https://github.com/moby/moby/tree/master/client functionality. WDYT?

simonaronsson replied:
I definitely agree!

If anything, I think it could end up being too much. I don't want to assume too much docker knowledge from the user. But let's start from here. If this extension takes off and people seem to be confused with how to use the props for some reason, we could always limit it in a future version.

dgzlopes reviewed (commented):
I'm excited about this extension :) 

Left some comments (for @simme primarily). Also, I think you should vendor your deps!

simme commented:
I'm _guessing_ you meant to mention @simskij :)

dgzlopes commented:
Oh, sorry @simme. I can't trust GH auto-complete :smile:

simonaronsson commented on docker.go:
```diff
@@ -0,0 +1,44 @@
+}
+
+// Ps lists Docker containers
+func (d *Docker) Ps() []string {
```
Should this be renamed to `ListContainers` or similar? I'm thinking that there might be no need to have matching terminology if it makes the module harder to read and use. 👍🏼

lxkuz replied:
✅

simskij commented:
Great work! 👏🏼  Added a few comments.

lxkuz committed:
Rework ListContainers method

lxkuz committed:
Add StartContainer method

lxkuz committed:
Add StopContainer method

lxkuz committed:
Move docker ps, stop and start under k6/x/docker/containers namespace

lxkuz committed:
Extract containers logic into separate file

lxkuz committed:
Add PauseContainer method

lxkuz committed:
Add UnpauseContainer method

lxkuz committed:
Add Docker logs method

lxkuz committed:
Pull client setup to separate method

lxkuz committed:
Add Docker inspect method

lxkuz committed:
Add Docker exec method

lxkuz committed:
Polish examples

lxkuz commented:
@simskij @dgzlopes Ok I'm removing WIP label as my PR becomes huge and I would like to wrap it up and continue work in tiny PRs. I reworked it since your last review and added the list of docker commands.
What issues I got:
1) `docker run` seems more difficult, will cover it in the next PR.
2) I did not vendor my deps, please let me know how should I do it properly.
3) `docker logs` works weird, see the example. I guess it connected with the buffer somehow, I don't have an exact explanation.

dgzlopes commented:
Yup! If we've smaller PRs, it's easier to review them. 

To vendor the dependencies, usually, I run `go mod tidy` and then `go mod vendor`. Once you've done that, you can add and commit the vendor folder.

lxkuz committed:
Vendor dependencies

simonaronsson commented on containers.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+// SetupClient for filling in Docker client instance
+func (d *Containers) SetupClient() {
+	cli, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		panic(err)
+	}
+	d.Client = cli
+}
```
Maybe we should allow the user to configure this by supplying a config object as well and not strictly enforce config through env vars?

I'm thinking that if an object is supplied as the argument to the function, we should treat that as a client config object, if not - default to `FromEnv`. If we decide to go for this, it could be good if we also cover it with a unit test for each of the paths.

lxkuz replied:
So you mean the usage will look like this?
```js
import containers from 'k6/x/docker/containers';

export default function () {
  const containerId = 'YOUR_CONTAINER_ID';
  const env = {
    ....
  }
  containers.setup(env)
  containers.startContainer(containerId, {attach: true});
}

```

simonaronsson replied:
Yeah, exactly! 👍🏼

simonaronsson commented on Readme.md:
plain might be preferable here given the weird syntax highlighting of the output?

lxkuz replied:
✅

simonaronsson commented on containers.go:
```diff
@@ -0,0 +1,81 @@
+
+	return d.Client.ContainerStop(context.Background(), containerID, &timeout)
+}
+
+// PauseContainer works as Docker pause command
+func (d *Containers) PauseContainer(containerID string) error {
+	return d.Client.ContainerPause(context.Background(), containerID)
+}
+
+// UnpauseContainer works as Docker unpause command
+func (d *Containers) UnpauseContainer(containerID string) error {
+	return d.Client.ContainerUnpause(context.Background(), containerID)
+}
+
+// Logs works as Docker logs command
+func (d *Containers) Logs(containerID string, options types.ContainerLogsOptions) (string, error) {
+	response, err := d.Client.ContainerLogs(context.Background(), containerID, options)
+	buf := new(bytes.Buffer)
+
+	if err != nil {
+		return "", err
+	}
+
+	buf.ReadFrom(response)
+	return buf.String(), nil
+}
+
+// InspectContainer works as Docker inspect command
+func (d *Containers) InspectContainer(containerID string) (types.ContainerJSON, error) {
+	return d.Client.ContainerInspect(context.Background(), containerID)
+}
+
+// ExecContainer works as Docker exec command
+func (d *Containers) ExecContainer(containerID string, config types.ExecConfig) error {
+	response, err := d.Client.ContainerExecCreate(context.Background(), containerID, config)
+	if err != nil {
+		return err
+	}
+	return d.Client.ContainerExecStart(context.Background(), response.ID, types.ExecStartCheck{})
+}
```
Now that I think about it, and see it in the context of the other commands, maybe we should omit `Container` from the method names? 

`containers.listContainers` and `containers.unpauseContainer` looks a bit excessive in comparison to `containers.list` and `containers.unpause`, don't you think? Since they are all available from the `containers` object, it should be quite obvious what they refer to anyhow.

lxkuz replied:
✅

simonaronsson commented on examples/docker-logs.js:
```diff
@@ -0,0 +1,19 @@
+    tail: 20,
+    // defaults: true
+  }
+  const result = containers.logs(containerId, options);
```
This goes for all of the functions that take an options object: Should we provide some "sane" defaults in this case if a user omits the options object from the call? Or is the docker client already taking care of that?

simonaronsson commented on examples/docker-pause.js:
```diff
@@ -0,0 +1,6 @@
+
+export default function () {
+  const containerId = 'YOUR_CONTAINER_ID';
+  containers.pauseContainer(containerId, {attach: true});
```
```suggestion
  containers.pauseContainer(containerId);
```

implementation does not take any args

lxkuz replied:
✅

simonaronsson commented on examples/docker-start.js:
```diff
@@ -0,0 +1,6 @@
+
+export default function () {
+  const containerId = 'YOUR_CONTAINER_ID';
+  containers.startContainer(containerId, {attach: true});
```
What would attaching even mean in this context? That we block until the container exits?

lxkuz replied:
This interface is wrong, I just remove it and keep containerId only

lxkuz replied:
✅

simonaronsson reviewed (changes requested):
Looks superb! That was really fast of you! 🎉 👍🏼
Left some minor suggestions.

simskij commented:
As discussed on slack, I'd also suggest having a look at whether we can generate typescript types for the extension using something like https://github.com/OneOfOne/struct2ts or https://github.com/tkrajina/typescriptify-golang-structs. This would definitely improve the developer experience while writing scripts, regardless if the user is using typescript or not.

lxkuz committed:
Exclude 'Container' word from container method names

lxkuz committed:
Cleanup containers Start

lxkuz committed:
Fix containers Pause example

lxkuz committed:
Polish Readme

simskij commented:
Yeah, I'll go ahead and merge this and then we can continue in smaller PR:s. 👍🏼
